src/storage/data_object_storage_registry.rs

```diff
@@ -0,0 +1,150 @@
+    }
+
+    fn is_ready_at_storage_provider(which: &T::ContentId, provider: &T::AccountId) -> bool {
+        let dosr_list = Self::data_object_storage_relationships(which);
+        return dosr_list.iter().any(|&dosr_id| {
+            let dosr = Self::data_object_storage_relationship(dosr_id).unwrap();
+            dosr.storage_provider == *provider && dosr.ready
+        });
+    }
+}
+
+
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event<T>() = default;
+
+        pub fn add_data_object_storage_relationship(origin, cid: T::ContentId) {
+            // Origin has to be a storage provider
+            let who = ensure_signed(origin).clone().unwrap();
+            // TODO check for being staked as a storage provider
+            // if !T::IsActiveMember::is_active_member(&who) {
+            //     return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            // }
+
+            // Content ID must exist
+            if !T::ContentIdExists::has_content(&cid) {
+                return Err(MSG_CID_NOT_FOUND);
+            }
+
+            // Create new ID, data.
+            let new_id = Self::next_data_object_storage_relationship_id();
+            let dosr: DataObjectStorageRelationship<T> = DataObjectStorageRelationship {
+                content_id: cid.clone(),
+                storage_provider: who.clone(),
+                ready: false,
+            };
+
+            <DataObjectStorageRelationshipMap<T>>::insert(new_id, dosr);
+            <NextDataObjectStorageRelationshipId<T>>::mutate(|n| { *n += T::DataObjectStorageRelationshipId::sa(1); });
+
+            // Also add the DOSR to the list of DOSRs for the CID. Uniqueness is guaranteed
+            // by the map, so we can just append the new_id to the list.
+            let mut dosr_list = Self::data_object_storage_relationships(cid.clone());
+            dosr_list.push(new_id);
+
+            // Emit event
+            Self::deposit_event(RawEvent::DataObjectStorageRelationshipAdded(new_id, cid, who));
+        }
+
+        // A storage provider may flip their own ready state, but nobody else.
+        pub fn set_data_object_storage_relationship_ready(origin, id: T::DataObjectStorageRelationshipId) {
+            Self::toggle_dosr_ready(origin, id, true)?;
+        }
+
+        pub fn unset_data_object_storage_relationship_ready(origin, id: T::DataObjectStorageRelationshipId) {
+            Self::toggle_dosr_ready(origin, id, false)?;
+        }
+    }
+}
+
+impl <T: Trait> Module<T> {
+    fn toggle_dosr_ready(origin: T::Origin, id: T::DataObjectStorageRelationshipId, ready: bool) -> dispatch::Result
+    {
+        // Origin has to be the storage provider mentioned in the DOSR
+        let who = ensure_signed(origin).clone().unwrap();
+
+        // For that, we need to fetch the identified DOSR
+        let found = Self::data_object_storage_relationship(id).ok_or(MSG_DOSR_NOT_FOUND);
```

Comment:
If you change the return type to the std Result type (not substrate dispatch::Result) this can be refactored to:
`let found = Self::data_object_storage_relationship(id).ok_or(MSG_DOSR_NOT_FOUND)?;`
and you wound't need to unwrap() on next line.

Reply:
Actually you can do it without needing to change the return type?

Reply:
Let me try that!